Arman Arkilic

#1 Implement Monte Carlo Algorithm as seen in class for European Call option

In [1]:
import numpy as np
import math

In [2]:
def european_call_mc(s_0, K, T, r, sigma, i):
    """European call option algorithm using Monte Carlo simulation"""
    # Get the random variables out of standard normal
    z = np.random.standard_normal(I)
    s_T = s_0 * np.exp((r - 0.5 * sigma**2) * T + sigma * np.sqrt(T) * z)
    h_t = np.maximum(s_T-K, 0)
    return np.exp(-r * T) * np.sum(h_t) / I

#2 Use your algorithm to compute the value at time 0 provided variables below:

In [3]:
s_0 = 120
K = 100
T = 1.0
r = 0.01
sigma = 0.2

I = 100000
z = np.random.standard_normal(I)

In [4]:
result = european_call_mc(s_0, K, T, r, sigma, I)

In [5]:
result

22.921443601243912

#3 Compare the outcome with the value obtained from the exact formula given in class:

In [6]:
s_0 = 120
K = 100
T = 1.0
r = 0.01
sigma = 0.2

I1 = int(1e5)
I2 = int(1e6)
z = np.random.standard_normal(I)

result100k = european_call_mc(s_0, K, T, r, sigma, I1)
result1mil = european_call_mc(s_0, K, T, r, sigma, I2)

res_dict = {'exp_1_mil': result1mil,
           'exp_100k': result100k}

The price of the 
$v(t,x) = sN(d_{+}(T −t,s))−exp(−r(T −t))KN(d_{-} (T −t,s))$ where

$d_{+}(\tau, s) = 1/(\sigma * sqrt(\tau))[log(s/K) + (r + \sigma ^2) \tau]$

$d_{+} = ( log(S/K)+(r+1/2*\sigma^2))/(\sigma*sqrt(T))$ = 0.21232 / 0.2 = 1.0615999999999999

$d_{-} = d_{+} - \sigma*sqrt(T)$ = 0.8616

$SN(d_{+}) - Ke^{-r\tau}N(d_{-})$ = 22.945710000000005

In [7]:
res_dict['formula'] = 22.945710000000005

In [8]:
print(res_dict)

{'exp_100k': 22.899805344698429, 'formula': 22.945710000000005, 'exp_1_mil': 22.899198265423053}


Clearly, with increasing number of iterations, the simulated outcome has less error compared to formula.

#4 Cash-or-nothing binary call option: if $S_T >_{=} K$ pay Q, otherwise pay 0

In [9]:
def indicator(s, k):
    """Provides a vectorized way to calculate binary indicator"""
    if not isinstance(s, np.ndarray):
        raise ValueError('S must be a numpy array')
    return s > k

def binary_call_mc(s_0, K, T, r, sigma, i, q):
    """European call option algorithm using Monte Carlo simulation"""
    # Get the random variables out of standard normal
    z = np.random.standard_normal(I)
    s_T = s_0 * np.exp((r - 0.5 * sigma**2) * T + sigma * np.sqrt(T) * z)
    h_t = indicator(s_T, K)
    return np.exp(-r * T) * np.sum(h_t) / I * q

#5

In [10]:
s_0 = 120
K = 100
T = 1.0
r = 0.01
sigma = 0.2
Q = 10

I = int(1e6)
z = np.random.standard_normal(I)

In [11]:
result = binary_call_mc(s_0, K, T, r, sigma, I, Q)

In [12]:
result

7.972703002710313

#6 Under risk neutral measure:

$S(t) = S(0) * e^{\widetilde{W} \sigma + (r - 1/2\sigma^2)t}$ then

$S(T) = S(t) * e^{(\widetilde{W}(T))-\widetilde{W}(t))\sigma + (r - 1/2\sigma^2)t}$

$Y = - \frac{(\widetilde{W}(T) -  \widetilde{W}(t))}{\sqrt(T-t)}$ and Y~N(0, 1)

By rewriting S(T) using Y, I get:

$S(T) = S(t) e^{-\sigma \sqrt{T-t}Y + (r-0.5\sigma^2)(T-t)}$

Then the option price can be written as:

$v(t, s) = \widetilde{E}[e^{-r(T-t)} Q X_{S(t)exp{-\sigma\sqrt{T-t}Y + (r-0.5\sigma^2)(T-t)}>K}$

$v(t, s) = \frac{1}{\sqrt{2\pi}} \int \limits_{-\infty}^{d_{-}} e^{-r (T-t)} Q exp(-0.5 y^2) dy$

$= \frac{Q}{\sqrt{2\pi}}e^{-r(T-t)}\int \limits_{-\infty}^{d_{-}}exp(-0.5 y^2) dy$

$= Q e^{-r(T-t)} N(d_{-})$  where $d_{-} = \frac{log(S/K) + (r - \sigma^2)(T-t)}{\sigma\sqrt{T-t}}$ and N is standard normal distro cdf!


#7 

$r=0.01, S_{0}=120, K=100, T=1, \sigma=0.2, Q=10, t=0$

Then $v(t,s) = v(.120)=10 e^{-0.01} N(0.8616) = 7.97$

#8 
By looking at the outcome of the MC simulation on binary option (from output [12], we can clearly state that the simulation approximates the option price well since error is about 0.0098 between the closed form solution and the simulation